README.md
# GitHub Update Checker

A Unity script that checks if a newer release of your Unity plugin is available on GitHub.

**This tool assumes your GitHub releases are tagged with a version number. eg. 1.0.0 or v1.0.0**

## Usage

This script is designed to be included in the distribution of your Unity plugin:

1.  Clone the repo into your Unity script directory. Delete the directory `DELETE_ME` as it contains some test stuff.
2.  Instantiate the checker:

        void OnEnable() {
            GitHub_Update_Checker githubUpdateChecker = new GitHub_Update_Checker() {
                githubOwner = "your_name",
                githubRepo = "your_repo_name",
                currentVersion = File.ReadAllText("Assets/myplugin/VERSION.txt", System.Text.Encoding.UTF8)
            };
        }

3.  Render the output:

        void OnGUI() {
            githubUpdateChecker.Render();
        }

Tested in editor windows but should work inside custom inspectors.

## Rate limiting

Auto-check is disabled by default because **GitHub rate limits your IP to 60 requests per hour**. If multiple Unity scripts use this plugin it will cause them all to break. Use that toggle wisely.

## Properties - Configuration

All of these properties are public so you can configure it as much as you want and so you can build your own checker.

\* = required

### string githubOwner\*

The owner of the repo.

### string githubRepo\*

The name of the repo to check.

### string currentVersion\*

The current version to use when checking for new versions.

Supports any version you can give to `System.Version` and can include a "v". eg. `1.0.0` or `v1.0.0`

### bool initialCheck: true

If to check for new versions when the script starts.

### bool autoCheck: false

If to automatically check for versions every minute (the maximum for rate-limiting).

### bool allowManualCheck: false

Allow the user to click a button to manually check for versions. Disabled by default to avoid rate-limiting issues.

### bool allowDownload: true

Allow the user to download the 1st asset of the release to their Downloads folder (`~/Downloads/owner_repo/myfile.asset`). If no binary is set, the sourcecode ZIP is downloaded.

### bool allowInstall: true

Allow the user to automatically install the asset (if it is a .unitypackage). The user always has the option of opening the containing folder.

## Properties - Other

### int autoCheckIntervalMs: 60000

The interval (in ms) for auto-checking. See rate-limiting section.

### States state

The state of the script - waiting, idle, checking for releases, downloading, error.

### Release availableRelease

A GitHub release if it is newer than our release (or null if none found).

### string downloadedReleaseAssetPath

The path to the asset that was downloaded (or null if none downloaded yet).

## Ideas for future

- detect if any assets have already been downloaded
- disable manual check button if it has been less than 1 minute to avoid rate-limiting


CHANGELOG.md
# 1.0.1

- changed all text to be regular size except "Update is available"
- fixed errors going into Play mode

# 1.0.0

- first version


Editor/DELETE_ME/GitHub_Update_Checker_TestWindow.cs

using UnityEditor;
using UnityEngine;

public class GitHub_Update_Checker_TestWindow : EditorWindow {
    GitHub_Update_Checker githubUpdateChecker;

    [MenuItem("PeanutTools/GitHub_Update_Checker_TestWindow")]
    public static void ShowWindow() {
        var window = GetWindow<GitHub_Update_Checker_TestWindow>();
        window.titleContent = new GUIContent("GitHub_Update_Checker_TestWindow");
        window.minSize = new Vector2(250, 50);
    }

    void OnEnable() {
        githubUpdateChecker = new GitHub_Update_Checker() {
            githubOwner = "imagitama",
            githubRepo = "webextension-boilerplate",
            currentVersion = "v1.0.0",
            allowManualCheck = true,
            allowInstall = false
        };
    }

    void OnGUI() {
        githubUpdateChecker.Render();
    }
}

Editor/GitHub_Update_Checker.cs
#pragma warning disable CS4014,CS1998
using UnityEditor;
using UnityEngine;

using System;
using System.IO;
using System.Collections;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using System.Runtime.InteropServices;
using System.Net.Http;
using Octokit;

public class GitHub_Update_Checker {
    public enum States {
        Waiting,
        Idle,
        Checking,
        Downloading,
        Error
    }

    private static readonly GitHubClient github = new GitHubClient(new ProductHeaderValue("GitHub_Update_Checker"));
    private static readonly int maximumSecondsBetweenRequests = 60; 

    // user settings
    public string githubOwner; // "imagitama"
    public string githubRepo; // "Unity-GitHub-Update-Checker"
    public string currentVersion; // "1.2.3"
    public bool initialCheck = true;
    public bool autoCheck = false; // disabled by default to prevent rate limit issues
    public bool allowManualCheck = false; // disabled by default to prevent rate limit issues
    public bool allowDownload = true;
    public bool allowInstall = true;

    public States state = States.Waiting;
    public int autoCheckIntervalMs = 60000; // warning: github rate limits your IP to 60 requests per hour
    public Octokit.Release availableRelease;
    public string downloadedReleaseAssetPath;

    public GitHub_Update_Checker() {
        if (initialCheck) {
            Debug.Log("Performing initial check...");

            CheckForAvailableReleaseWithDelay();
        } else {
            Debug.Log("Skipping initial check");
        }

        if (autoCheck) {
            StartAutoChecking();
        } else {
            Debug.Log("Skipping auto-check");
        }
    }

    int GetNow() {
        return (int)DateTimeOffset.Now.ToUnixTimeSeconds();
    }

    int GetTimestampOfLastCheck() {
        return EditorPrefs.GetInt("GitHub_Update_Checker_LastCheck", GetNow());
    }

    int GetSecondsSinceLastCheck() {
        return GetNow() - GetTimestampOfLastCheck();
    }

    void UpdateTimestampOfLastCheck() {
        EditorPrefs.SetInt("GitHub_Update_Checker_LastCheck", GetNow());
    }

    async Task<IReadOnlyList<Octokit.Release>> GetAllReleases() {   
        Debug.Log("Getting all releases...");

        var releases = await github.Repository.Release.GetAll("imagitama", "vrc-questifyer");

        Debug.Log("Found " + releases.Count + " releases");

        return releases;
    }

    async Task StartAutoChecking() {
        Debug.Log("Starting auto-checking every " + autoCheckIntervalMs + "ms...");

        await CheckForAvailableReleaseWithDelay();

        while (autoCheck == true) {
            await Task.Delay(autoCheckIntervalMs);

            if (state != States.Downloading) {
                await CheckForAvailableReleaseWithDelay();
            }
        }
    }

    public string GetVersionNumber(string versionStr) {
        if (versionStr.Substring(0, 1).ToLower() == "v") {
            return versionStr.Substring(1);
        }
        return versionStr;
    }

    public async Task CheckForAvailableReleaseWithDelay() {
        try {
            int secondsSinceLastCheck = GetSecondsSinceLastCheck();

            if (secondsSinceLastCheck < maximumSecondsBetweenRequests) {
                int secondsToWait = maximumSecondsBetweenRequests - secondsSinceLastCheck;

                Debug.Log("It has been " + secondsSinceLastCheck + " seconds since our last check, waiting " + secondsToWait + " seconds...");
                
                await Task.Delay(secondsToWait * 1000);
                
                Debug.Log("Wait complete, checking...");
            }
            
            await CheckForAvailableRelease();
        } catch (Exception exception) {
            state = States.Error;
            HandleException(exception);
        }
    }

    public async Task CheckForAvailableRelease() {
        try {
            state = States.Checking;

            UpdateTimestampOfLastCheck();

            var releases = await GetAllReleases();
            Version bestVersionSoFar = new Version(GetVersionNumber(currentVersion));
            Octokit.Release possibleAvailableRelease = null;

            foreach (var release in releases) {
                if (release.TagName == null || release.TagName == "") {
                    throw new Exception("Release does not have a tag name!");
                }

                string thisVersionNumber = GetVersionNumber(release.TagName);
                Version thisVersion = new Version(thisVersionNumber);

                if (thisVersion.CompareTo(bestVersionSoFar) == 1) {
                    bestVersionSoFar = thisVersion;
                    possibleAvailableRelease = release;
                }
            }

            if (possibleAvailableRelease != null) {
                Debug.Log("Version " + bestVersionSoFar.ToString() + " is available");

                availableRelease = possibleAvailableRelease;
            }
            
            state = States.Idle;
        } catch (Exception exception) {
            state = States.Error;
            HandleException(exception);
        }
        
        ForceRefreshGUI();
    }

    void ManuallyCheckForAvailableRelease() {
        int secondsSinceLastCheck = GetSecondsSinceLastCheck();

        if (secondsSinceLastCheck < maximumSecondsBetweenRequests) {
            Debug.LogWarning("You are manually checking even though it has only been " + secondsSinceLastCheck.ToString() + " seconds since last check!");
        }

        CheckForAvailableRelease();
    }

    void HandleException(Exception exception) {
        Debug.LogError(exception);
    }

    // source: https://stackoverflow.com/a/7672816/1215393
    public static string GetDownloadsPath() {
        SHGetKnownFolderPath(new Guid("{374DE290-123F-4565-9164-39C4925E467B}"), 0x00004000, new IntPtr(0), out var PathPointer);
        var Result = Marshal.PtrToStringUni(PathPointer);
        Marshal.FreeCoTaskMem(PathPointer);
        return Result;
    }

    Octokit.ReleaseAsset GetBestReleaseAsset(Octokit.Release release) {
        if (release.Assets.Count >= 1) {
            return release.Assets[0];
        } else {
            return null;
        }
    }

    [DllImport("Shell32.dll")]
    static extern int SHGetKnownFolderPath([MarshalAs(UnmanagedType.LPStruct)] Guid rfid, uint dwFlags, IntPtr hToken, out IntPtr ppszPath);

    public async Task DownloadAvailableRelease() {
        try {
            state = States.Downloading;

            Debug.Log("Downloading available release...");

            Debug.Log("There are " + availableRelease.Assets.Count + " assets for release \"" + availableRelease.Name + "\"");

            Octokit.ReleaseAsset bestAsset = GetBestReleaseAsset(availableRelease);

            string downloadUrl = "";
            string downloadPath = GetDownloadsPath();

            if (bestAsset != null) {
                Debug.Log("Downloading the best asset \"" + bestAsset.Name + "\"...");

                downloadUrl = bestAsset.BrowserDownloadUrl;
                downloadPath += "\\" + bestAsset.Name;
            } else {
                Debug.Log("Best asset not found, using source code zip...");

                downloadUrl = availableRelease.ZipballUrl;
                downloadPath += "\\" + availableRelease.TagName + ".zip";
            }

            if (File.Exists(downloadPath)) {
                Debug.Log("File already exists (" + downloadPath + "), deleting...");

                File.Delete(downloadPath);
            }

            HttpClient client = new HttpClient();
            client.DefaultRequestHeaders.Add("Accept", "application/octet-stream");

            var response = await client.GetAsync(downloadUrl);

            Debug.Log("Destination: " + downloadPath);

            using (var fs = new FileStream(downloadPath, System.IO.FileMode.CreateNew)) {
                await response.Content.CopyToAsync(fs);
            }

            downloadedReleaseAssetPath = downloadPath;

            Debug.Log("File has been downloaded");

            state = States.Idle;

            ForceRefreshGUI();
        } catch (Exception exception) {
            HandleException(exception);
        }
    }

    void ForceRefreshGUI() {
        GUI.FocusControl(null);
    }

    public async Task InstallDownloadedAsset() {
        System.Diagnostics.Process.Start(downloadedReleaseAssetPath);
    }

    bool CanAssetBeInstalled(string pathToAsset) {
        return pathToAsset.Contains(".unitypackage");
    }

    bool RenderButton(string label) {
        return GUILayout.Button(label, GUILayout.Width(150), GUILayout.Height(50));
    }

    void RenderText(string text) {
        var labelStyle = new GUIStyle(GUI.skin.label) {
            alignment = TextAnchor.UpperCenter
        };

        GUILayout.Label(text, labelStyle);
    }

    void RenderLargeText(string text) {
        var labelStyle = new GUIStyle(GUI.skin.label) {
            fontSize = 16,
            alignment = TextAnchor.UpperCenter
        };
        GUILayout.Label(text, labelStyle);
    }

    void RenderDisabledText(string text) {
        var labelStyle = new GUIStyle(GUI.skin.label) {
            alignment = TextAnchor.UpperCenter
        };
        labelStyle.normal.textColor = new Color(1.0f, 1.0f, 1.0f, 0.5f);

        GUILayout.Label(text, labelStyle);
    }

    void RenderGap() {
        EditorGUILayout.Space();
    }

    void OpenInExplorer(string itemPath) {
        itemPath = itemPath.Replace(@"/", @"\");
        System.Diagnostics.Process.Start("explorer.exe", "/select," + itemPath);
    }

    void OpenUrl(string url) {
        UnityEngine.Application.OpenURL(url);
    }

    public void Render() {
        if (githubRepo == null || currentVersion == null) {
            return;
        }

        switch (state) {
            case States.Waiting:
                RenderDisabledText("Waiting to check for updates...");
                return;
            case States.Checking:
                RenderText("Checking for updates...");
                RenderGap();
                break;
            case States.Error:
                RenderDisabledText("Failed to check for updates");
                RenderGap();
                break;
            case States.Downloading:
                RenderText("Downloading...");
                RenderGap();
                break;
            default:
                if (availableRelease != null) {
                    RenderLargeText("A newer version is available: " + availableRelease.TagName);

                    RenderGap();

                    if (downloadedReleaseAssetPath != null) {
                        RenderText("Package has been downloaded");

                        RenderGap();

                        if (allowInstall && CanAssetBeInstalled(downloadedReleaseAssetPath)) {
                            if (RenderButton("Install")) {
                                InstallDownloadedAsset();
                                
                                RenderGap();
                            }
                        }

                        if (RenderButton("Open Folder")) {
                            OpenInExplorer(downloadedReleaseAssetPath);
                        }
                    } else {
                        if (allowDownload) {
                            if (RenderButton("Download")) {
                                DownloadAvailableRelease();
                                
                                RenderGap();
                            }
                        }

                        if (RenderButton("View on GitHub")) {
                            OpenUrl("https://www.github.com/" + githubOwner + "/" + githubRepo + "/releases/tag/" + availableRelease.TagName);
                        }
                    }
                } else {
                    RenderText("No new version is available");
                }
            break;
        }

        if (state == States.Idle && allowManualCheck) {
            RenderGap();

            if (RenderButton("Check For Updates")) {
                ManuallyCheckForAvailableRelease();
            }
        }
    }
}